### PR TITLE
Fix CRDs deletion in operator deletion

### PR DESCRIPTION
**Pull request by arnongilboa**
Also check DataImportCron CRD has no DeletionTimestamp before adding a
finalizer

Signed-off-by: Arnon Gilboa <agilboa@redhat.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: pkg/operator/controller/controller_test.go [hunk 1]
```diff
@@ -1523,12 +1523,14 @@ func createArgs() *args {
 }
 
 func doReconcile(args *args) {
-	result, err := args.reconciler.Reconcile(context.TODO(), reconcileRequest(args.cdi.Name))
-	Expect(err).ToNot(HaveOccurred())
-	Expect(result.Requeue).To(BeFalse())
+	Eventually(func() bool {
+		result, err := args.reconciler.Reconcile(context.TODO(), reconcileRequest(args.cdi.Name))
+		return err == nil && !result.Requeue
+	}, 1*time.Minute, 2*time.Second).Should(BeTrue())
```
Comment (mhenriks) [inline pkg/operator/controller/controller_test.go:1529]: What's going on here?  Unit tests should be deterministic.

File: pkg/operator/controller/callbacks.go [hunk 2]
```diff
@@ -132,6 +107,28 @@ func reconcileDeleteControllerDeployment(args *callbacks.ReconcileCallbackArgs)
...
 
+func deleteCRDs(args *callbacks.ReconcileCallbackArgs) error {
+	crdNames := []string{
+		"dataimportcrons.cdi.kubevirt.io",
+		"datavolumes.cdi.kubevirt.io",
+		"objecttransfers.cdi.kubevirt.io",
+	}
+	crdsExist := false
+	for _, crdName := range crdNames {
+		crd := &extv1.CustomResourceDefinition{ObjectMeta: metav1.ObjectMeta{Name: crdName}}
```
Comment (mhenriks) [inline pkg/operator/controller/callbacks.go:118]: Maybe changing this to the following to address unit test issues?

Get CRD if not exist continue
If CRD.deletionTimestamp == nil{
  Delete CRD
  Get CRD if not exist continue
}
crdsExist = true

File: pkg/operator/controller/controller_test.go (hunk 1 above)
Comment (mhenriks) [reply pkg/operator/controller/controller_test.go:1529]: I think I get it

File: pkg/operator/controller/callbacks.go (hunk 2 above)
Comment (arnongilboa) [reply pkg/operator/controller/callbacks.go:118]: So the issue was caused by repeated Delete() of CRD which already had DeletionTimestamp? Late night games...
Fixed

File: pkg/operator/controller/controller_test.go (hunk 1 above)
Comment (arnongilboa) [reply pkg/operator/controller/controller_test.go:1529]: Thanks. Reverted it.

File: pkg/operator/controller/callbacks.go [hunk 3]
```diff
@@ -132,6 +107,44 @@ func reconcileDeleteControllerDeployment(args *callbacks.ReconcileCallbackArgs)
...
+			}
+			if err := args.Client.Get(context.TODO(), types.NamespacedName{Name: crdName}, crd); err != nil {
+				if errors.IsNotFound(err) {
+					continue
+				}
+				return err
+			}
+			args.Logger.Info("CRD is not deleted yet", "crdName", crdName)
+			crdsExist = true
+		}
```
Comment (mhenriks) [inline pkg/operator/controller/callbacks.go:140]: you have to set `crdsExist = true` here is else block when `crd.DeletionTimestamp != nil`

Comment (arnongilboa) [reply pkg/operator/controller/callbacks.go:138]: Fat fingers. Fixed.

File: pkg/controller/dataimportcron-controller.go [hunk 4]
```diff
@@ -120,7 +121,19 @@ func (r *DataImportCronReconciler) Reconcile(ctx context.Context, req reconcile.
 }
 
 func (r *DataImportCronReconciler) initCron(ctx context.Context, dataImportCron *cdiv1.DataImportCron) error {
-	AddFinalizer(dataImportCron, dataImportCronFinalizer)
+	log := r.log.WithName("initCron")
+	if !HasFinalizer(dataImportCron, dataImportCronFinalizer) {
+		crd := &extv1.CustomResourceDefinition{}
+		if err := r.client.Get(context.TODO(), types.NamespacedName{Name: "dataimportcrons.cdi.kubevirt.io"}, crd); err == nil {
```
Comment (mhenriks) [inline pkg/controller/dataimportcron-controller.go:127]: err == nil
Don’t think that’s correct

File: pkg/controller/dataimportcron-controller.go [hunk 5]
```diff
@@ -120,7 +121,19 @@ func (r *DataImportCronReconciler) Reconcile(ctx context.Context, req reconcile.
...
+			log.Error(err, "Cannot get CRD")
+			return nil
```
Comment (mhenriks) [inline pkg/controller/dataimportcron-controller.go:129]: Return err?

File: pkg/controller/dataimportcron-controller.go [hunk 6]
```diff
@@ -120,7 +121,19 @@ func (r *DataImportCronReconciler) Reconcile(ctx context.Context, req reconcile.
...
+		}
+		if crd.DeletionTimestamp != nil {
+			log.Info("CRD has DeletionTimestamp")
+			return nil
```
Comment (mhenriks) [inline pkg/controller/dataimportcron-controller.go:133]: Return an error instead maybe?

File: pkg/controller/dataimportcron-controller.go (hunk 4 above)
Comment (arnongilboa) [reply pkg/controller/dataimportcron-controller.go]: fixed

**Comment by mhenriks** [comment]
new func test may be flaky

**Comment by arnongilboa** [comment]
> new func test may be flaky

@mhenriks it was not the new func test, but a missing `extv1.AddToScheme()` in the controller for the CRD. Added.

**Pull request merged**